packages/cli/src/commands/testCommand.ts

```diff
@@ -15,18 +15,47 @@
  */
 
 import { Command } from 'commander';
-import { run } from 'helpers/run';
 import { paths } from 'helpers/paths';
+import { runCheck } from 'helpers/run';
+
+function includesAnyOf(hayStack: string[], ...needles: string[]) {
+  for (const needle of needles) {
+    if (hayStack.includes(needle)) {
+      return true;
+    }
+  }
+  return false;
+}
 
 export default async (cmd: Command) => {
-  const args = ['test', '--config', paths.resolveOwn('config/jest.js')];
+  // all args are forwarded to jest
+  const rawArgs = cmd.parent.rawArgs as string[];
+  const args = rawArgs.slice(rawArgs.indexOf('test') + 1);
 
-  if (cmd.watch) {
-    args.push('--watch');
+  // Only include our config if caller isn't passing their own config
+  if (!includesAnyOf(args, '-c', '--config')) {
+    args.push('--config', paths.resolveOwn('config/jest.js'));
   }
-  if (cmd.coverage) {
-    args.push('--coverage');
+
+  // Run in watch mode unless in CI, coverage mode, or running all tests
+  if (
+    !process.env.CI &&
+    !args.includes('--coverage') &&
+    // explicitly no watching
+    !includesAnyOf(args, '--no-watch', '--watch=false', '--watchAll=false') &&
+    // already watching
+    !includesAnyOf(args, '--watch', '--watchAll')
+  ) {
+    const isGitRepo = () => runCheck('git rev-parse --is-inside-work-tree');
+    const isMercurialRepo = () => runCheck('hg --cwd . root');
```

Comment:
These look like important helper functions, which could be re-used in future. IMO, we should organize them as helper functions somewhere. Or at least, move them on the top of the module.

I am also thinking that this will make it easier to test these, and the newly created `includesAnyOf` function above. :) 🙇

Reply:
I'd say move them out if there's more usage, not too happy about utilities only used in one place :p. I do agree it's a bit more likely that this would be used else, but there's already other logic for checking whether we're inside git that does it in a different way that I don't think is worth refactoring [here](https://github.com/spotify/backstage/blob/e842489333af3653032edeff382a63addf936b0c/packages/cli/src/commands/build-cache/cache.ts#L89).

As for testing I think we're gonna have to rely more on e2e tests for these command wrappers. Although it might make sense to add some just to increase friction of changing things a bit, forcing you to think a bit more about the change.

Reply:
> As for testing I think we're gonna have to rely more on e2e tests for these command wrappers.

Agreed 💯

Reply:
Does this handle gracefully the case when the hg command does not exist?

Reply:
@freben Yeah error or non-0 exit code just returns false